robyn/testing.py

```diff
@@ -0,0 +1,155 @@
+    
+    def __init__(self, queries = {}, headers = {}, path_params = {}, method = "GET", ip_addr = None):
+        self.queries = queries
+        self.headers = headers
+        self.path_params = path_params
+        self.method = method
+        self.ip_addr = ip_addr
+
+class TestClient:
+    #helper function
+    def get_route(self, path, method):
+        routes = self.app.router.get_routes()
+        r = None
+        for route in routes: 
+            if route.route == path and route.route_type == method:
+                r = route
+                break
+        return r
+    
+    def __init__(self, app):
+        self.app = app
+    
+    def get(self, method_path, **kwargs):
+        return self.do_test_request(HttpMethod.GET, method_path, **kwargs)
+    
+    def post(self, method_path, **kwargs):
+        return self.do_test_request(HttpMethod.POST, method_path, **kwargs)
+    
+    def delete(self, method_path, **kwargs):
+        return self.do_test_request(HttpMethod.DELETE, method_path, **kwargs)
+    
+    def patch(self, method_path, **kwargs):
+        return self.do_test_request(HttpMethod.PATCH, method_path, **kwargs)
+        
+    def options(self, method_path, **kwargs):
+        return self.do_test_request(HttpMethod.OPTIONS, method_path, **kwargs)
+        
+    def head(self, method_path, **kwargs):
+        return self.do_test_request(HttpMethod.HEAD, method_path, **kwargs)
+        
+    def trace(self, method_path, **kwargs):
+        return self.do_test_request(HttpMethod.TRACE, method_path, **kwargs)
+        
+    def connect(self, method_path, **kwargs):
+        return self.do_test_request(HttpMethod.CONNECT, method_path, **kwargs)
+    
+    def do_test_request(self, method, method_path, data=None, json=None, headers=None, files=None, params=None):
+        route = self.get_route(method_path, method)
+        if route == None:
+            return None
+        req = TestRequest(method = method)
+        #default headers
+        req.headers["host"] = "localhost"
+        req.headers["connection"] = "keep-alive"
+        req.headers["user-agent"] = "robyn-testclient"
+        req.headers["accept"] = "*/*"
+        req.headers["accept-encoding"] = "gzip, deflate"
+        #input headers
+        if headers != None:
+            for header in headers:
+                req.headers[header] = headers[header]
+        #input parameters
+        if params != None:
+            if type(params) == dict:
+                req.queries = params
+            elif type(params) == list:
+                for param in params:
+                    req.queries[param[0]] = param[1]
+        #input body
+        if files != None:
+            body, header = encode_multipart_formdata(files)
+            req.headers["content-type"] = header
+            req.body = list(body)
+            req.headers["content-length"] = len(req.body)
+        elif data != None:
+            if type(data) == bytes:
```

Comment:
I don't think we need to do for single type there might a function in python that would be handelling it